### PR TITLE
Fix formatter handling of blank lines

### LINKED ISSUE #862: Policy formatter drops empty lines in multi-line strings
### Bug Category

Other

### Describe the bug

Given a policy

```cedar
permit (principal, action, resource)
when { "

a
" };
```

The formatter drops the new-line character and rewrites the policy to

```cedar
permit (principal, action, resource)
when { "
a
" };
```

This is probably because the [`RcDoc::as_string`](https://docs.rs/pretty/latest/pretty/struct.RcDoc.html#method.as_string) function cannot be used with strings containing line breaks. Possibly related to #265.

### Reproduction steps

```console
[jkastner@dev-dsk-jkastner-1a-3309db3b cedar]$ cargo run format <<<'
permit (principal, action, resource)
when { "

a"};'

permit (principal, action, resource)
when { "
a" };
```

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

The cause of #862 was the call to `remove_empty_lines`, which removed empty lines from the formatted policy -- including those in string literals.

_Attempt 1:_ It was simple enough to remove the `remove_empty_lines` call, but this makes formatted policies ugly since the current code introduces extra newlines around comments. I imagine the reason `remove_empty_lines` was originally introduced was to handle these extra newlines.

_Attempt 2:_ I tried to get rid of the extra newlines by adjusting how `RcDoc`s are constructed, but that ended up being a pain in the butt.

_Attempt 3 (current PR):_ I added a new function `remove_empty_lines_safe`, which does some light parsing to avoid removing newlines in strings.

### Summary

* Added the `remove_empty_lines_safe` function **← the important change**
* Switched to using `RcDoc::text` in cases where the call to `to_string` was unnecessary
* Added comments in some parts of the code that are subtle.

The last two changes are just edits I made while getting familiar with the codebase. I think they're harmless.

We had previously discussed not using `RcDoc::text` or `RcDoc::as_string` for string parsing. But I think it’s actually ok: newlines are [used in the tests](https://docs.rs/pretty/latest/src/pretty/lib.rs.html#1771). I think the way to interpret the "must not contain line breaks" guidance is that indentation will not be handled correctly when you do this, which is actually what we want for string literals.

## Issue #, if available

Resolves #862

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy-formatter/src/pprint/utils.rs [hunk 1]
```diff
@@ -24,27 +24,27 @@ pub fn add_brackets<'a>(d: RcDoc<'a>, leftp: RcDoc<'a>, rightp: RcDoc<'a>) -> Rc
...
+    trailing_comment: &str,
+    next_doc: RcDoc<'a>,
+) -> RcDoc<'a> {
     if trailing_comment.is_empty() {
-        RcDoc::nil()
+        next_doc
     } else {
         RcDoc::space()
-            .append(RcDoc::text(trailing_comment.trim().to_owned()))
+            .append(create_multiline_doc(trailing_comment))
```
Comment (john-h-kastner-aws) [inline cedar-policy-formatter/src/pprint/utils.rs:47]: I'm surprised that this doesn't causes changes in any test cases. Also, I'm not sure what impact, if any, this change actually should have

File: cedar-policy-formatter/src/pprint/fmt.rs [hunk 2]
```diff
@@ -43,7 +42,8 @@ fn tree_to_pretty<T: Doc>(t: &T, context: &mut config::Context<'_>) -> Result<St
 }
 
 fn soundness_check(ps: &str, ast: &PolicySet) -> Result<()> {
-    let formatted_ast = parse_policyset(ps).wrap_err("formatter produces invalid policies")?;
+    let formatted_ast =
+        parse_policyset(ps).wrap_err(format!("formatter produced an invalid policy set:\n{ps}"))?;
```
Comment (john-h-kastner-aws) [inline cedar-policy-formatter/src/pprint/fmt.rs:46]: While we're changing this text, we should make it clear that this is and internal error should be reported

**Comment by khieta** [comment]
Marking as ready-for-review since the code is in a reviewable state & CI is passing. I plan on running the formatter fuzzing targets later today -- will add a comment here with the results.

File: cedar-policy-formatter/src/pprint/utils.rs [hunk 3]
```diff
@@ -112,26 +126,73 @@ pub fn get_comment_in_range(span: miette::SourceSpan, tokens: &mut [WrappedToken
...
+    // PANIC SAFETY: this regex pattern is valid
+    #[allow(clippy::unwrap_used)]
+    let string_regex = Regex::new(r#""(\\.|[^"\\])*"[^\n]*\n"#).unwrap();
+
+    let mut index = 0;
+    let mut final_text = String::new();
+
+    while index < text.len() {
+        // Check for the next comment and string. The general strategy is to
+        // call `remove_empty_lines` on all the text _outside_ of comments and
```
Comment (john-h-kastner-aws) [inline cedar-policy-formatter/src/pprint/utils.rs:167]: I found this comment little confusing. IIUC, it doesn't particularly matter if we remove empty lines in comments. Comments can't be empty lines. Skipping past comments is necessary in this loop to properly handling something like `// comment "not a string`. Would be good to call out this and any other edge cases you handled.

**Review by john-h-kastner-aws — APPROVED** [review]
Look pretty good.

File: cedar-policy-formatter/src/pprint/utils.rs [hunk 4]
```diff
@@ -112,26 +126,73 @@ pub fn get_comment_in_range(span: miette::SourceSpan, tokens: &mut [WrappedToken
...
+}
+
+/// Remove empty lines, safely handling newlines that occur in quotations.
+pub fn remove_empty_lines_safe(text: &str) -> String {
+    // PANIC SAFETY: this regex pattern is valid
+    #[allow(clippy::unwrap_used)]
+    let comment_regex = Regex::new(r"//[^\n]*\n").unwrap();
+    // PANIC SAFETY: this regex pattern is valid
+    #[allow(clippy::unwrap_used)]
+    let string_regex = Regex::new(r#""(\\.|[^"\\])*"[^\n]*\n"#).unwrap();
```
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy-formatter/src/pprint/utils.rs]: Allowing escape characters like `\"`

File: cedar-policy-formatter/src/pprint/utils.rs [hunk 5]
```diff
@@ -112,26 +126,73 @@ pub fn get_comment_in_range(span: miette::SourceSpan, tokens: &mut [WrappedToken
...
+    while index < text.len() {
+        // Check for the next comment and string. The general strategy is to
+        // call `remove_empty_lines` on all the text _outside_ of comments and
+        // strings.
+        let comment_match = comment_regex.find_at(text, index);
+        let string_match = string_regex.find_at(text, index);
+        match (comment_match, string_match) {
+            (Some(m1), Some(m2)) => {
+                // Handle the earlier match
+                let m = if m1.start() < m2.start() { m1 } else { m2 };
```
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy-formatter/src/pprint/utils.rs:184]: I don't want the minimum start index though, I want the match with the smaller start. Is there some special syntax for that?

**Comment by khieta** [comment]
Ran the `formatter` target for an hour with no issue. Ran the `formatter-bytes` target for a few minutes and it immediately found the issue (nice addition @john-h-kastner-aws!). I've fixed the identified bug, and now I'm letting the `formatter-bytes` target run some more 🙂

File: cedar-policy-formatter/src/pprint/utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-formatter/src/pprint/utils.rs:184]: `min_by_key` might work

**Review by shaobo-he-aws — APPROVED** [review]
LGTM. Thanks a lot for fixing it.

File: cedar-policy-formatter/src/pprint/utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy-formatter/src/pprint/utils.rs]: Updated in the latest commit

**Review by john-h-kastner-aws — APPROVED** [review]

**Comment by khieta** [comment]
Update: `formatter-bytes` has been running for an hour and a half with no failures, so I'm going to call this good.

**Pull request merged**